File: src/python/xerxes/tools/computer_use/cua_backend.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Cua-driver backend (macOS only).

Speaks MCP over stdio to `cua-driver`. The Python `mcp` SDK is async, so we
run a dedicated asyncio event loop on a background thread and marshal sync
calls through it.

Install: `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/trycua/cua/main/libs/cua-driver/scripts/install.sh)"`

After install, `cua-driver` is on $PATH and supports `cua-driver mcp` (stdio
transport) which is what we invoke.

The private SkyLight SPIs cua-driver uses (SLEventPostToPid, SLPSPostEvent-
RecordTo, _AXObserverAddNotificationAndCheckRemote) are not Apple-public and
can break on OS updates. Pin the installed version via `XERXES_CUA_DRIVER_
VERSION` if you want reproducibility across an OS bump.
"""

from __future__ import annotations

import asyncio
import base64
import logging
import os
import re
import shutil
import sys
import threading
from typing import Any

from .backend import (
    ActionResult,
    CaptureResult,
    ComputerUseBackend,
    UIElement,
)

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Version pinning
# ---------------------------------------------------------------------------

PINNED_CUA_DRIVER_VERSION = os.environ.get("XERXES_CUA_DRIVER_VERSION", "0.5.0")

_CUA_DRIVER_CMD = os.environ.get("XERXES_CUA_DRIVER_CMD", "cua-driver")
_CUA_DRIVER_ARGS = ["mcp"]  # stdio MCP transport

# Regex to parse list_windows text output lines:
#   "- AppName (pid 12345) "Title" [window_id: 67890]"
_WINDOW_LINE_RE = re.compile(
    r'^-\s+(.+?)\s+\(pid\s+(\d+)\)\s+.*\[window_id:\s+(\d+)\]',
    re.MULTILINE,
)

# Regex to parse element lines from get_window_state AX tree markdown.
#
# Handles two output formats from different cua-driver versions:
#   Classic:  "  - [N] AXRole \"label\""
#   New:       "[N] AXRole (order) id=Label"
#
# Group 1: element index
# Group 2: AX role
# Group 3: quoted label (classic format)
# Group 4: id= label (new format)
_ELEMENT_LINE_RE = re.compile(
    r'^\s*(?:-\s+)?\[(\d+)\]\s+(\w+)(?:\s+"([^"]*)"|(?:\s+\(\d+\))?\s+id=([^\s\[\]]*))?' ,
    re.MULTILINE,
)


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _is_macos() -> bool:
    return sys.platform == "darwin"


def cua_driver_binary_available() -> bool:
    """True if `cua-driver` is on $PATH or XERXES_CUA_DRIVER_CMD resolves."""
    return bool(shutil.which(_CUA_DRIVER_CMD))


def cua_driver_install_hint() -> str:
    return (
        "cua-driver is not installed. Install with:\n"
        "  /bin/bash -c \"$(curl -fsSL "
        "https://raw.githubusercontent.com/trycua/cua/main/libs/cua-driver/scripts/install.sh)\"\n"
        "Or set XERXES_CUA_DRIVER_CMD to the binary path."
    )


def _parse_windows_from_text(text: str) -> list[dict[str, Any]]:
    """Parse window records from list_windows text output."""
    windows = []
    for m in _WINDOW_LINE_RE.finditer(text):
        windows.append({
            "app": m.group(1).strip(),
            "pid": int(m.group(2)),
            "window_id": int(m.group(3)),
        })
    return windows


def _parse_elements_from_text(text: str) -> list[UIElement]:
    """Parse AX element records from get_window_state markdown."""
    elements = []
    for m in _ELEMENT_LINE_RE.finditer(text):
        idx = int(m.group(1))
        role = m.group(2)
        label = (m.group(3) or m.group(4) or "").strip()
        elements.append(UIElement(index=idx, role=role, label=label))
    return elements


def _b64_from_png_bytes(png_bytes: bytes) -> str:
    return base64.b64encode(png_bytes).decode("ascii")


# ---------------------------------------------------------------------------
# CuaBackend
# ---------------------------------------------------------------------------

class CuaBackend(ComputerUseBackend):
    """Sync wrapper around the async cua-driver MCP client.

    A private asyncio event loop runs on a background thread so that every
    public method can be called synchronously from the agent's tool executor.
    """

    def __init__(self) -> None:
        self._loop: asyncio.AbstractEventLoop | None = None
        self._thread: threading.Thread | None = None
        self._session: Any = None  # mcp.ClientSession
        self._tools: list[dict[str, Any]] = []
        self._started = False

    # ── Lifecycle ───────────────────────────────────────────────────

    def start(self) -> None:
        if self._started:
            return
        if not _is_macos():
            raise RuntimeError("CuaBackend is macOS-only")
        if not cua_driver_binary_available():
            raise RuntimeError(cua_driver_install_hint())

        self._loop = asyncio.new_event_loop()
        self._thread = threading.Thread(target=self._loop.run_forever, daemon=True)
        self._thread.start()

        # Connect to cua-driver via MCP stdio
        future = asyncio.run_coroutine_threadsafe(self._connect(), self._loop)
        future.result(timeout=15)
        self._started = True

    def stop(self) -> None:
        if not self._started or self._loop is None:
            return
        future = asyncio.run_coroutine_threadsafe(self._disconnect(), self._loop)
        try:
            future.result(timeout=5)
        except Exception:
            pass
        self._loop.call_soon_threadsafe(self._loop.stop)
        if self._thread:
            self._thread.join(timeout=2)
        self._started = False

    def is_available(self) -> bool:
        return _is_macos() and cua_driver_binary_available()

    # ── Internal async plumbing ─────────────────────────────────────

    async def _connect(self) -> None:
        """Spawn cua-driver and initialize the MCP session."""
        try:
            from mcp import ClientSession, StdioServerParameters
            from mcp.client.stdio import stdio_client
        except ImportError as exc:
            raise RuntimeError(
                "mcp SDK not installed. Run: uv pip install 'mcp>=1.0.0'"
            ) from exc

        server_params = StdioServerParameters(
            command=_CUA_DRIVER_CMD,
            args=_CUA_DRIVER_ARGS,
            env=None,
        )

        self._stdio_ctx = stdio_client(server_params)
        read, write = await self._stdio_ctx.__aenter__()

        self._session = ClientSession(read, write)
        await self._session.__aenter__()
        await self._session.initialize()

        # Discover available tools
        tools_result = await self._session.list_tools()
        self._tools = [
            {"name": t.name, "description": t.description}
            for t in (tools_result.tools if hasattr(tools_result, "tools") else [])
        ]
        logger.debug("cua-driver tools: %s", [t["name"] for t in self._tools])

    async def _disconnect(self) -> None:
        if self._session:
            await self._session.__aexit__(None, None, None)
            self._session = None
        if hasattr(self, "_stdio_ctx"):
            await self._stdio_ctx.__aexit__(None, None, None)

    async def _call_tool(self, name: str, arguments: dict[str, Any]) -> dict[str, Any]:
        """Call a cua-driver tool by name and return the parsed result."""
        if self._session is None:
            raise RuntimeError("CuaBackend not started")

        result = await self._session.call_tool(name, arguments=arguments)

        # Parse content — can be text or image
        content = result.content if hasattr(result, "content") else []
        text_parts = []
        image_parts = []

        for part in content:
            if hasattr(part, "text"):
                text_parts.append(part.text)
            elif hasattr(part, "image_url"):
                image_parts.append(part.image_url)
            elif hasattr(part, "data") and hasattr(part, "mimeType"):
                # Binary image data
                if "png" in part.mimeType:
                    image_parts.append(_b64_from_png_bytes(part.data))

        return {
            "text": "\n".join(text_parts),
            "images": image_parts,
            "is_error": getattr(result, "isError", False),
        }

    # ── Public sync API ───────────────────────────────────────────

    def _run(self, coro: asyncio.Coroutine) -> Any:
        """Marshal an async call onto the background loop."""
        if self._loop is None:
            raise RuntimeError("CuaBackend not started")
        future = asyncio.run_coroutine_threadsafe(coro, self._loop)
        return future.result(timeout=30)

    # ── Capture ─────────────────────────────────────────────────────

    def capture(self, mode: str = "som", app: str | None = None) -> CaptureResult:
        args: dict[str, Any] = {"mode": mode}
        if app:
            args["app"] = app

        result = self._run(self._call_tool("capture", args))
        text = result.get("text", "")
        images = result.get("images", [])

        png_b64 = images[0] if images else None

        # Parse elements from text response
        elements = _parse_elements_from_text(text)

        # Try to infer dimensions from the image if available
        width, height = 0, 0
        if png_b64:
            try:
                png_bytes = base64.b64decode(png_b64)
                # PNG header: width at bytes 16-20, height at 20-24
                if len(png_bytes) >= 24:
                    width = int.from_bytes(png_bytes[16:20], "big")
                    height = int.from_bytes(png_bytes[20:24], "big")
            except Exception:
                pass

        return CaptureResult(
            mode=mode,
            width=width,
            height=height,
            png_b64=png_b64,
            elements=elements,
            png_bytes_len=len(base64.b64decode(png_b64)) if png_b64 else 0,
        )

    # ── Pointer actions ─────────────────────────────────────────────

    def click(
        self,
        *,
        element: int | None = None,
        x: int | None = None,
        y: int | None = None,
        button: str = "left",
        click_count: int = 1,
        capture_after: bool = False,
    ) -> ActionResult:
        args: dict[str, Any] = {"button": button, "click_count": click_count}
        if element is not None:
            args["element"] = element
        elif x is not None and y is not None:
            args["x"] = x
            args["y"] = y
        else:
            return ActionResult(ok=False, action="click", message="Need element or x,y")

        if capture_after:
            args["capture_after"] = True

        result = self._run(self._call_tool("click", args))
        return ActionResult(
            ok=not result.get("is_error", False),
            action="click",
            message=result.get("text", ""),
        )

    def double_click(
        self,
        *,
        element: int | None = None,
        x: int | None = None,
        y: int | None = None,
        capture_after: bool = False,
    ) -> ActionResult:
        return self.click(
            element=element,
            x=x,
            y=y,
            button="left",
            click_count=2,
            capture_after=capture_after,
        )

    def right_click(
        self,
        *,
        element: int | None = None,
        x: int | None = None,
        y: int | None = None,
        capture_after: bool = False,
    ) -> ActionResult:
        return self.click(
            element=element,
            x=x,
            y=y,
            button="right",
            click_count=1,
            capture_after=capture_after,
        )

    def middle_click(
        self,
        *,
        element: int | None = None,
        x: int | None = None,
        y: int | None = None,
        capture_after: bool = False,
    ) -> ActionResult:
        return self.click(
            element=element,
            x=x,
            y=y,
            button="middle",
            click_count=1,
            capture_after=capture_after,
        )

    def drag(
        self,
        *,
        start_element: int | None = None,
        start_x: int | None = None,
        start_y: int | None = None,
        end_element: int | None = None,
        end_x: int | None = None,
        end_y: int | None = None,
        capture_after: bool = False,
    ) -> ActionResult:
        args: dict[str, Any] = {}
        if start_element is not None:
            args["start_element"] = start_element
        elif start_x is not None and start_y is not None:
            args["start_x"] = start_x
            args["start_y"] = start_y
        else:
            return ActionResult(ok=False, action="drag", message="Need start_element or start_x,start_y")

        if end_element is not None:
            args["end_element"] = end_element
        elif end_x is not None and end_y is not None:
            args["end_x"] = end_x
            args["end_y"] = end_y
        else:
            return ActionResult(ok=False, action="drag", message="Need end_element or end_x,end_y")

        if capture_after:
            args["capture_after"] = True

        result = self._run(self._call_tool("drag", args))
        return ActionResult(
            ok=not result.get("is_error", False),
            action="drag",
            message=result.get("text", ""),
        )

    def scroll(
        self,
        *,
        element: int | None = None,
        x: int | None = None,
        y: int | None = None,
        dx: int = 0,
        dy: int = 0,
        capture_after: bool = False,
    ) -> ActionResult:
        args: dict[str, Any] = {"dx": dx, "dy": dy}
        if element is not None:
            args["element"] = element
        elif x is not None and y is not None:
            args["x"] = x
            args["y"] = y

        if capture_after:
            args["capture_after"] = True

        result = self._run(self._call_tool("scroll", args))
        return ActionResult(
            ok=not result.get("is_error", False),
            action="scroll",
            message=result.get("text", ""),
        )

    # ── Keyboard actions ────────────────────────────────────────────

    def type(self, text: str, capture_after: bool = False) -> ActionResult:
        args: dict[str, Any] = {"text": text}
        if capture_after:
            args["capture_after"] = True
        result = self._run(self._call_tool("type", args))
        return ActionResult(
            ok=not result.get("is_error", False),
            action="type",
            message=result.get("text", ""),
        )

    def key(self, key: str, capture_after: bool = False) -> ActionResult:
        args: dict[str, Any] = {"key": key}
        if capture_after:
            args["capture_after"] = True
        result = self._run(self._call_tool("key", args))
        return ActionResult(
            ok=not result.get("is_error", False),
            action="key",
            message=result.get("text", ""),
        )

    def set_value(self, value: str, element: int | None = None, capture_after: bool = False) -> ActionResult:
        args: dict[str, Any] = {"value": value}
        if element is not None:
            args["element"] = element
        if capture_after:
            args["capture_after"] = True
        result = self._run(self._call_tool("set_value", args))
        return ActionResult(
            ok=not result.get("is_error", False),
            action="set_value",
            message=result.get("text", ""),
        )

    # ── Wait / App management ─────────────────────────────────────

    def wait(self, ms: int = 1000) -> ActionResult:
        result = self._run(self._call_tool("wait", {"ms": ms}))
        return ActionResult(
            ok=not result.get("is_error", False),
            action="wait",
            message=result.get("text", ""),
        )

    def list_apps(self) -> ActionResult:
        result = self._run(self._call_tool("list_apps", {}))
        text = result.get("text", "")
        windows = _parse_windows_from_text(text)
        return ActionResult(
            ok=not result.get("is_error", False),
            action="list_apps",
            message=text,
            meta={"windows": windows},
        )

    def focus_app(self, app: str) -> ActionResult:
        result = self._run(self._call_tool("focus_app", {"app": app}))
        return ActionResult(
            ok=not result.get("is_error", False),
            action="focus_app",
            message=result.get("text", ""),
        )

File: src/typescript/src/tools/computerUse/cuaBackend.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { Buffer } from 'node:buffer'

import { ClientError } from '../../core/errors.js'
import { MCPClient } from '../../mcp/client.js'
import type { MCPToolCallResult } from '../../mcp/types.js'
import type { JsonObject } from '../../types/toolCalls.js'
import {
  ComputerUseUnavailableError,
  type ActionResult,
  type CaptureRequest,
  type CaptureResult,
  type ClickRequest,
  type ComputerUseBackend,
  type DragRequest,
  type KeyRequest,
  type ScrollRequest,
  type SetValueRequest,
  type TextRequest,
  type UIElement,
} from './backend.js'

const CUA_DRIVER_INSTALL_HINT = [
  'No cua-driver backend is configured.',
  'Install and run cua-driver behind an explicitly configured MCP transport, then inject CuaBackend into ComputerUseSession.',
  'The Bun runtime never downloads or auto-installs desktop automation software.',
].join(' ')

const WINDOW_LINE = /^-\s+(.+?)\s+\(pid\s+(\d+)\)\s+.*\[window_id:\s+(\d+)\]/gm
const ELEMENT_LINE = /^\s*(?:-\s+)?\[(\d+)\]\s+(\w+)(?:\s+"([^"]*)"|(?:\s+\(\d+\))?\s+id=([^\s\[\]]*))?/gm

/**
 * Narrow MCP boundary required by the CUA adapter.
 *
 * It is deliberately injected. This module does not inspect PATH, launch a
 * shell command, or install cua-driver. A host can use MCPClient below after
 * it has deliberately supplied a server configuration.
 */
export interface CuaDriverMcpPort {
  callTool(name: string, arguments_: JsonObject, signal?: AbortSignal): MCPToolCallResult | Promise<MCPToolCallResult>
  isAvailable(signal?: AbortSignal): boolean | Promise<boolean>
  start?(signal?: AbortSignal): void | Promise<void>
  stop?(signal?: AbortSignal): void | Promise<void>
}

export interface MCPClientCuaDriverPortOptions {
  /**
   * Host-defined provision check. It must not perform a privileged desktop
   * action. The default only means that the caller deliberately configured an
   * MCP client; actual connection failures still make start fail explicitly.
   */
  readonly isProvisioned?: (signal?: AbortSignal) => boolean | Promise<boolean>
}

/**
 * Adapter for the repository's MCP client after a host has chosen its process
 * command and security policy. It is not registered or constructed implicitly.
 */
export class MCPClientCuaDriverPort implements CuaDriverMcpPort {
  constructor(
    private readonly client: Pick<MCPClient, 'callTool' | 'connect' | 'disconnect'>,
    private readonly options: MCPClientCuaDriverPortOptions = {},
  ) {}

  async isAvailable(signal?: AbortSignal): Promise<boolean> {
    throwIfAborted(signal)
    return this.options.isProvisioned === undefined ? true : Boolean(await this.options.isProvisioned(signal))
  }

  async start(signal?: AbortSignal): Promise<void> {
    throwIfAborted(signal)
    await this.client.connect()
  }

  async stop(signal?: AbortSignal): Promise<void> {
    throwIfAborted(signal)
    await this.client.disconnect()
  }

  async callTool(name: string, arguments_: JsonObject, signal?: AbortSignal): Promise<MCPToolCallResult> {
    throwIfAborted(signal)
    return this.client.callTool(name, arguments_)
  }
}

/**
 * CUA-driver implementation of the generic computer-use backend.
 *
 * It faithfully maps the driver MCP surface into normalized desktop actions,
 * but owns no process launch policy. Pair it with an injected CuaDriverMcpPort
 * and a ComputerUseSession to make desktop access available to a tool call.
 */
export class CuaBackend implements ComputerUseBackend {
  private started = false

  constructor(private readonly driver: CuaDriverMcpPort) {}

  async start(signal?: AbortSignal): Promise<void> {
    if (this.started) return
    if (!await this.isAvailable(signal)) {
      throw new ComputerUseUnavailableError(cuaDriverInstallHint())
    }
    await this.driver.start?.(signal)
    this.started = true
  }

  async stop(signal?: AbortSignal): Promise<void> {
    if (!this.started) return
    try {
      await this.driver.stop?.(signal)
    } finally {
      this.started = false
    }
  }

  async isAvailable(signal?: AbortSignal): Promise<boolean> {
    throwIfAborted(signal)
    return Boolean(await this.driver.isAvailable(signal))
  }

  async capture(request: CaptureRequest, signal?: AbortSignal): Promise<CaptureResult> {
    const arguments_: JsonObject = { mode: request.mode }
    if (request.app !== undefined) arguments_.app = request.app
    const result = await this.call('capture', arguments_, signal)
    const pngB64 = firstPng(result)
    const dimensions = pngB64 === undefined ? { width: 0, height: 0 } : pngDimensions(pngB64)
    return {
      mode: request.mode,
      width: dimensions.width,
      height: dimensions.height,
      ...(pngB64 === undefined ? {} : { pngB64, pngBytesLength: decodedLength(pngB64) }),
      elements: parseCuaElements(textFromResult(result)),
    }
  }

  async click(request: ClickRequest, signal?: AbortSignal): Promise<ActionResult> {
    return this.clickWithAction('click', request, signal)
  }

  async doubleClick(request: Omit<ClickRequest, 'button' | 'clickCount'>, signal?: AbortSignal): Promise<ActionResult> {
    return this.clickWithAction('double_click', { ...request, button: 'left', clickCount: 2 }, signal)
  }

  async rightClick(request: Omit<ClickRequest, 'button' | 'clickCount'>, signal?: AbortSignal): Promise<ActionResult> {
    return this.clickWithAction('right_click', { ...request, button: 'right', clickCount: 1 }, signal)
  }

  async middleClick(request: Omit<ClickRequest, 'button' | 'clickCount'>, signal?: AbortSignal): Promise<ActionResult> {
    return this.clickWithAction('middle_click', { ...request, button: 'middle', clickCount: 1 }, signal)
  }

  async drag(request: DragRequest, signal?: AbortSignal): Promise<ActionResult> {
    const arguments_: JsonObject = captureArguments(request.captureAfter)
    assignOptionalNumber(arguments_, 'start_element', request.startElement)
    assignOptionalNumber(arguments_, 'start_x', request.startX)
    assignOptionalNumber(arguments_, 'start_y', request.startY)
    assignOptionalNumber(arguments_, 'end_element', request.endElement)
    assignOptionalNumber(arguments_, 'end_x', request.endX)
    assignOptionalNumber(arguments_, 'end_y', request.endY)
    return this.actionFromResult('drag', await this.call('drag', arguments_, signal))
  }

  async scroll(request: ScrollRequest, signal?: AbortSignal): Promise<ActionResult> {
    const arguments_: JsonObject = { ...captureArguments(request.captureAfter), dx: request.dx, dy: request.dy }
    assignOptionalNumber(arguments_, 'element', request.element)
    assignOptionalNumber(arguments_, 'x', request.x)
    assignOptionalNumber(arguments_, 'y', request.y)
    return this.actionFromResult('scroll', await this.call('scroll', arguments_, signal))
  }

  async type(request: TextRequest, signal?: AbortSignal): Promise<ActionResult> {
    return this.actionFromResult(
      'type',
      await this.call('type', { ...captureArguments(request.captureAfter), text: request.text }, signal),
    )
  }

  async key(request: KeyRequest, signal?: AbortSignal): Promise<ActionResult> {
    return this.actionFromResult(
      'key',
      await this.call('key', { ...captureArguments(request.captureAfter), key: request.key }, signal),
    )
  }

  async setValue(request: SetValueRequest, signal?: AbortSignal): Promise<ActionResult> {
    const arguments_: JsonObject = { ...captureArguments(request.captureAfter), value: request.value }
    assignOptionalNumber(arguments_, 'element', request.element)
    return this.actionFromResult('set_value', await this.call('set_value', arguments_, signal))
  }

  async wait(ms: number, signal?: AbortSignal): Promise<ActionResult> {
    return this.actionFromResult('wait', await this.call('wait', { ms }, signal))
  }

  async listApps(signal?: AbortSignal): Promise<ActionResult> {
    const result = await this.call('list_apps', {}, signal)
    const message = textFromResult(result)
    return {
      ok: !result.isError,
      action: 'list_apps',
      message,
      meta: { windows: parseCuaWindows(message) },
    }
  }

  async focusApp(app: string, signal?: AbortSignal): Promise<ActionResult> {
    return this.actionFromResult('focus_app', await this.call('focus_app', { app }, signal))
  }

  private async clickWithAction(action: string, request: ClickRequest, signal?: AbortSignal): Promise<ActionResult> {
    const arguments_: JsonObject = {
      ...captureArguments(request.captureAfter),
      button: request.button,
      click_count: request.clickCount,
    }
    assignOptionalNumber(arguments_, 'element', request.element)
    assignOptionalNumber(arguments_, 'x', request.x)
    assignOptionalNumber(arguments_, 'y', request.y)
    return this.actionFromResult(action, await this.call('click', arguments_, signal))
  }

  private async call(name: string, arguments_: JsonObject, signal: AbortSignal | undefined): Promise<MCPToolCallResult> {
    throwIfAborted(signal)
    if (!this.started) {
      throw new ComputerUseUnavailableError('CuaBackend has not been started; use it through ComputerUseSession')
    }
    return this.driver.callTool(name, arguments_, signal)
  }

  private actionFromResult(action: string, result: MCPToolCallResult): ActionResult {
    return {
      ok: !result.isError,
      action,
      message: textFromResult(result),
    }
  }
}

/** Human-readable setup guidance only. It never attempts to install software. */
export function cuaDriverInstallHint(): string {
  return CUA_DRIVER_INSTALL_HINT
}

/** Parse the CUA driver's list_windows markdown into JSON-safe window records. */
export function parseCuaWindows(text: string): JsonObject[] {
  const windows: JsonObject[] = []
  WINDOW_LINE.lastIndex = 0
  let match: RegExpExecArray | null
  while ((match = WINDOW_LINE.exec(text)) !== null) {
    const app = match[1]?.trim()
    const pid = parsePositiveInteger(match[2])
    const windowId = parsePositiveInteger(match[3])
    if (app && pid !== undefined && windowId !== undefined) {
      windows.push({ app, pid, window_id: windowId })
    }
  }
  return windows
}

/** Parse the CUA driver's accessibility-tree markdown into generic UI elements. */
export function parseCuaElements(text: string): UIElement[] {
  const elements: UIElement[] = []
  ELEMENT_LINE.lastIndex = 0
  let match: RegExpExecArray | null
  while ((match = ELEMENT_LINE.exec(text)) !== null) {
    const index = parsePositiveInteger(match[1])
    const role = match[2]?.trim()
    if (index === undefined || !role) continue
    const label = (match[3] ?? match[4] ?? '').trim()
    elements.push(label ? { index, role, label } : { index, role })
  }
  return elements
}

/** Decode PNG IHDR dimensions when a cua-driver response includes a valid header. */
export function pngDimensions(pngB64: string): { readonly height: number; readonly width: number } {
  try {
    const bytes = Buffer.from(pngB64, 'base64')
    if (bytes.length < 24) return { width: 0, height: 0 }
    return {
      width: bytes.readUInt32BE(16),
      height: bytes.readUInt32BE(20),
    }
  } catch {
    return { width: 0, height: 0 }
  }
}

function captureArguments(captureAfter: boolean): JsonObject {
  return captureAfter ? { capture_after: true } : {}
}

function assignOptionalNumber(target: JsonObject, key: string, value: number | undefined): void {
  if (value !== undefined) target[key] = value
}

function textFromResult(result: MCPToolCallResult): string {
  return result.content
    .filter((content): content is Extract<MCPToolCallResult['content'][number], { readonly type: 'text' }> => content.type === 'text')
    .map(content => content.text)
    .join('\n')
}

function firstPng(result: MCPToolCallResult): string | undefined {
  const image = result.content.find(
    (content): content is Extract<MCPToolCallResult['content'][number], { readonly type: 'image' }> =>
      content.type === 'image' && content.mimeType.toLowerCase().includes('png'),
  )
  return image?.data
}

function decodedLength(base64: string): number {
  return Buffer.byteLength(base64, 'base64')
}

function parsePositiveInteger(value: string | undefined): number | undefined {
  if (value === undefined || !/^\d+$/.test(value)) return undefined
  const number = Number(value)
  return Number.isSafeInteger(number) && number > 0 ? number : undefined
}

function throwIfAborted(signal: AbortSignal | undefined): void {
  if (signal?.aborted) {
    throw new ClientError('computer_use', 'computer-use action was cancelled before the backend call')
  }
}
